### PR TITLE
More generic GUI widget structure

### LINKED ISSUE #1291: Visualizer: Can't drag World View when Visualizer window very small
### Description of the task

When the Visualizer is running but in a very small window, the user is unable to click and drag around the World view, despite some field being out of view.

### Acceptance criteria

- [ ] Allow the user to drag around the World view no matter the window size

### LINKED ISSUE #1565: Split up GUI widget further
### Description of the task

In order to help organize a better GUI structure, we should try make more of our "custom widget" full independent widgets. Rather than being regular widgets added in the `.ui` files with custom `Setup` functions, tney should be their own class entirely in the `generic_widgets` folder. There they can define their own behavior and expose any inteface they need to. Then these widgets can easily be re-used by adding them to any `.ui` file as a "promoted widget".

This will help compartmentalize widget more and make them easier to change and re-use.

This can be done for
* DynamicParameters
* Drawing the field / any scene using draw functions
* Showing PlayInfo

### Acceptance criteria

- [ ] Make a separate DynamicParameters widget
- [ ] Make a separate DrawFunction widget
- [ ] Make a separate PlayInfo widget

### LINKED ISSUE #1566: GUI: Allow "scaled resize" of field drawing
### Description of the task

The GUI window can't be shrunk any smaller than the field being rendered, making it very hard to use on small screens.

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

This PR further beaks up GUI components into their own re-usable widgets. This makes it easier to define custom behavior, control widget interfaces, and re-use custom widgets in different GUIs via the `.ui` files.

Note: This PR contains no functional changes other than the minor bugfixes listed below. All GUIs look and behave effectively the same. These changes just make it much easier to implement future changes.

Things I fixed along the way:
* `argc` and `argv` are now mocked, so we don't need to provide them to the GUI constructors
* Bugs where the field view couldn't be moved / shrunk properly

A few new conventions to note (these will be added to the documentation in #1567 if people are ok with them)
* I combined the `MainWidget` classes with the `...GUI` classes that used to contain them. They were really just another level of abstraction that passed data through, but didn't do anything useful
* If we ever need custom functionality from anything, implement it in a custom widget class
* `.ui` files should only be used to organize all the layout and widgets in an application. They should not set up signals/slots or any "logic"
* The top-level GUI classes should just aggregate the widgets they need and connect any callbacks / logic between them

This should help give a more clear idea of the "division of responsibility" between custom widgets, `.ui` files, and the top-level GUI classes (the actual GUI applications) 

### Testing Done

Visual testing

### Resolved Issues

resolves #1565 
resolves #1291 
resolves #1566

### Length Justification

One thing led to another and it was easier to do all at once, sorry. Also wanted this to go in quickly together to make more changes before Saturday.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/gui/drawing/draw_functions.h [hunk 1]
```diff
@@ -14,54 +16,97 @@
...
  *
  * DrawFunctions can capture any data that supports copying and
- * "store" it for later. That way the entire DrawFunction can be copied
+ * "store" it for later. That way the entire DrawFunctionWrapper can be copied
  * and passed around the system, until some consumer calls it to
  * draw its contents. This allows us to draw larger non-copyable objects
- * by creating a DrawFunction from it's copyable components / members.
+ * by creating a DrawFunctionWrapper from it's copyable components / members.
+ *
+ * This class is made abstract by making the desctructor pure virtual,
```
Comment (garethellis0) [inline src/software/gui/drawing/draw_functions.h]:
``` suggestion
 * This class is made abstract by making the destructor pure virtual,
```

File: src/software/gui/full_system/widgets/full_system_gui.cpp [hunk 2]
```diff
@@ -11,23 +12,44 @@ FullSystemGUI::FullSystemGUI(
...
+    // clicked. Specifically, we do this so that when the user clicks outside of the
+    // QLineEdits used for Parameters, the QLineEdit will lose focus.
+    // https://www.qtcentre.org/threads/41128-Need-to-implement-in-place-line-edit-unable-to-get-lose-focus-of-QLineEdit
+    setFocusPolicy(Qt::StrongFocus);
+
+    setupRobotStatusTable(main_widget->robot_status_table_widget);
+    setupAIControls(main_widget, config);
+
+    // Update to make sure all layout changes apply nicely
+    update();
```
Comment (garethellis0) [inline src/software/gui/full_system/widgets/full_system_gui.cpp]: If we didn't call this here, wouldn't it get called in the timer callback anyhow?

File: src/software/simulation/standalone_simulator_main.cpp [hunk 3]
```diff
@@ -1,26 +1,29 @@
 #include "software/gui/standalone_simulator/threaded_standalone_simulator_gui.h"
 #include "software/logger/logger.h"
-#include "software/parameter/dynamic_parameters.h"
-#include "software/simulation/standalone_simulator.h"
+#include "software/simulation/threaded_simulator.h"
+#include "software/world/field.h"
 
-int main(int argc, char* argv[])
+int main(int argc, char *argv[])
```
Comment (garethellis0) [inline src/software/simulation/standalone_simulator_main.cpp]: Could we remove the args here if we're not using them?

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/software/gui/full_system/widgets/full_system_gui.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/software/gui/full_system/widgets/full_system_gui.cpp]: yea turns out we don't really need it. Removed here and in other GUI classes.

File: src/software/simulation/standalone_simulator_main.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/software/simulation/standalone_simulator_main.cpp]: TIL you don't need argc argv. Removed

File: src/software/gui/full_system/threaded_full_system_gui.cpp [hunk 4]
```diff
@@ -46,8 +46,14 @@ ThreadedFullSystemGUI::~ThreadedFullSystemGUI()
...
 }
 
-void ThreadedFullSystemGUI::createAndRunFullSystemGUI(int argc, char** argv)
+void ThreadedFullSystemGUI::createAndRunFullSystemGUI()
 {
+    // We mock empty argc and argv since they don't affect the behaviour of the GUI.
+    // This way we don't need to pass them all the way down from the start of the
+    // program
+    char* argv[] = {NULL};
+    int argc     = sizeof(argv) / sizeof(char*) - 1;
```
Comment (jonathanlew) [inline src/software/gui/full_system/threaded_full_system_gui.cpp:55]: are these being used anywhere anymore?

**Review by jonathanlew — COMMENTED** [review]
Not much to add, good cleanup

File: src/software/gui/full_system/threaded_full_system_gui.cpp (hunk 4 above)
Comment (MathewMacDougall) [reply src/software/gui/full_system/threaded_full_system_gui.cpp:55]: Yes, used to create the `QApplication` in this thread.

**Review by jonathanlew — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]

**Review by kelvinkoon — APPROVED** [review]

**Pull request merged**